Mapping for Short,Byte,Date,Time,Timestamp,Boolean,BigDecimal in Java Client

harshit-gangal opened the pull request:
<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/youtube/vitess/1495)
<!-- Reviewable:end -->

enisoc commented on java/client/src/main/java/com/youtube/vitess/client/Proto.java:
```diff
@@ -203,6 +210,26 @@ public EntityId apply(Map.Entry<byte[], ?> entry) {
+        this.value = ByteString.copyFromUtf8(((boolean)value) ? "1" : "0");
+      } else if (value instanceof BigDecimal) {
+        // BigDecimal
+        this.type = Query.Type.FLOAT64;
```
I think BigDecimal should map to Query.Type.DECIMAL. Also when converting to string, I think BigDecimal.toPlainString() should work better than using setScale().

harshit-gangal replied:
The problem with that is the precision can change.
Ex. BigDecimal(0.23), BigDecimal.toPlanString() will result in 0.2300000000000000099920072216264088638126850128173828125
In MySQL the maximum scale is 30.
So if a column is defined as my_col decimal(2,30) then this will result in wrong data in the database.

enisoc replied:
I see, so you use setScale() to limit to 30. Is that really necessary? In my tests, MySQL will do the rounding for you already; it doesn't mind if the value is longer than will fit.

The reason I think we need toPlainString() instead of toString() is because toString() will revert to exponential notation if the exponent is < -6. That is, `new BigDecimal("0.0000000002").toString()` will become "2E-10". That's a problem because MySQL will interpret "2E-10" as an approximate, floating point numeric literal, rather than as an exact numeric literal:

http://dev.mysql.com/doc/refman/5.6/en/precision-math-numbers.html

To make sure MySQL interprets it as an exact numeric literal, we need to send "0.0000000002", which we can get from toPlainString().

harshit-gangal replied:
In MySQL you can have a scale of atmost 30.
So, If one gives value of BigDecimal as 0.23 and I store is as toPlainString then the value saved in database will be "0.230000000000000009992007221626" (Scale 30) and not "0.23" so on retrieval, they will get wrong value

sougou commented on java/client/src/main/java/com/youtube/vitess/client/Proto.java:
```diff
@@ -203,6 +210,26 @@ public EntityId apply(Map.Entry<byte[], ?> entry) {
+      } else if (value instanceof Boolean ) {
+        // Boolean
+        this.type = Query.Type.BIT;
+        this.value = ByteString.copyFromUtf8(((boolean)value) ? "1" : "0");
```
The string here needs to be a binary where the right-most bit must be set. See this test https://github.com/youtube/vitess/blob/master/go/vt/tabletserver/endtoend/compatibility_test.go#L331.
I actually don't know too much about bit field behavior in MySQL. I'll extend that test to check a few more value combinations.

enisoc replied:
If we could get Vitess to pass "0" or "1" to MySQL unquoted, this would work (I tried it).

@sougou If we use Query.Type.UINT64 (BIT fields are limited to 64 bits anyway), would Vitess pass the numbers unquoted?

sougou replied:
Yes. This should work too.

harshit-gangal replied:
boolean field is true or false so it has to be converted to 1 or 0.

As suggested I can use Query.Type.UINT64 for Boolean type but still conversion is required.

sougou commented:
Shouldn't there be counterpart changes to the place where results are received? For things to be symmetrical, those values need to be converted back to the native types.

enisoc commented:
We're already converting these types on the [receiving side](https://github.com/youtube/vitess/blob/ddac90d9f7df7b8b2dc88dfb2f56c57fb22af8bb/java/client/src/main/java/com/youtube/vitess/client/cursor/Row.java#L335), although we are using Joda DateTime instead of java.sql.Date and friends. We should think about whether it makes sense to switch the low-level Row class to return java.sql.Date, instead of making the JDBC wrapper convert from Joda to java.sql types.

enisoc commented:
I've proposed #1499 to fix up the receiving side to follow standard conventions better.

enisoc commented on java/client/src/main/java/com/youtube/vitess/client/Proto.java:
```diff
@@ -203,6 +210,26 @@ public EntityId apply(Map.Entry<byte[], ?> entry) {
+      } else if (value instanceof Timestamp) {
+        // DateTime, TimeStamp
+        this.type = Query.Type.TIMESTAMP;
+        this.value = ByteString.copyFromUtf8(value.toString());
```
I found out that toString() and valueOf() on java.sql.(Date|Time|Timestamp) all use the local machine's default timezone. It'll work out as long as users only ever create java.sql.Timestamp values via valueOf(). But what if they create one from a java.util.Date or `new Timestamp(millis)`? They would have to adjust their epoch times to be intentionally wrong to offset for the local machine's default timezone.

Ideally we would do everything in UTC, but that means we can use neither toString() nor valueOf(). We would have to do the parsing and formatting ourselves. I'm looking into it as part of #1499 where I'm working on the receiving side.

harshit-gangal replied:
This is known to Application Developer that is why they use SimpleDateFormat to generate date/time/timestamp field in the timezone they want and then send as setString() or setTimestamp(Timestamp.valueOf()) in the preparedStatement.

enisoc replied:
I see what you're saying. The default in JDBC when you don't specify a time zone, is to use the machine's configured time zone. Since we don't offer any way to specify time zone when sending bind variables as a map, we should follow the JDBC convention of using the machine's time zone.

FYI, in #1499 I'm adding support in Cursor for getting date/time values with a given Calendar. This should make implementing ResultSet time/date methods easier. There is also a new utility class in that PR called com.youtube.vitess.mysql.DateTime, which contains format methods that will also be useful when writing PreparedStatement setDate(..., Calendar) and so on.

I think to make that work, when implementing PreparedStatement, we'll need to send date/time bind vars to the low-level Vitess client as String, after formatting them with com.youtube.vitess.mysql.DateTime.

harshit-gangal replied:
That can be done.

enisoc commented:
Regarding BigDecimal, I think we're having trouble understanding each other's arguments. I've added a commit to this branch to show in code form what I'm proposing. Please see the new test case for BigDecimal that will fail if you use `toString()`. I've changed it to `toPlainString()` to show that the test passes. I kept the rounding when scale>30, but I don't think we should unnecessarily setScale(30) if the existing scale is small enough.

Lastly, regarding `new BigDecimal(0.23)`, the "error" is not introduced by `toPlainString()` as you implied. The error is in using the `BigDecimal(double)` constructor at all. That constructor is known to produce unexpected results:

https://docs.oracle.com/javase/7/docs/api/java/math/BigDecimal.html#BigDecimal(double)

If you instead use `BigDecimal.valueOf(double)`, the results will be as expected. I've made this change as well and added a test case to prove it works.

harshit-gangal commented:
I understand your point now.


On Thu, Feb 18, 2016 at 3:48 AM, Anthony Yeh <notifications@github.com>
wrote:

> Regarding BigDecimal, I think we're having trouble understanding each
> other's arguments. I've added a commit to this branch to show in code form
> what I'm proposing. Please see the new test case for BigDecimal that will
> fail if you use toString(). I've changed it to toPlainString() to show
> that the test passes. I kept the rounding when scale>30, but I don't think
> we should unnecessarily setScale(30) if the existing scale is small enough.
>
> Lastly, regarding new BigDecimal(0.23), the "error" is not introduced by
> toPlainString() as you implied. The error is in using the
> BigDecimal(double) constructor at all. That constructor is known to
> produce unexpected results:
>
>
> https://docs.oracle.com/javase/7/docs/api/java/math/BigDecimal.html#BigDecimal(double)
>
> If you instead use BigDecimal.valueOf(double), the results will be as
> expected. I've made this change as well and added a test case to prove it
> works.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/youtube/vitess/pull/1495#issuecomment-185432410>.
>

harshit-gangal commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[java/client/src/main/java/com/youtube/vitess/client/Proto.java, line 224 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1495#-KAhvODK6P7tr89-EAL1-r1-224:-KCdle-giAzDrkipNGrP:-499146009) ([raw file](https://github.com/youtube/vitess/blob/a8136b8bfe6ac7914062989903ea9e3667b65e19/java/client/src/main/java/com/youtube/vitess/client/Proto.java#L224)):*
Done.

---

*[java/client/src/main/java/com/youtube/vitess/client/Proto.java, line 228 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1495#-KAhvODK6P7tr89-EAL1-r1-228:-KCdljajvaGaC0oIKOY9:-499146009) ([raw file](https://github.com/youtube/vitess/blob/a8136b8bfe6ac7914062989903ea9e3667b65e19/java/client/src/main/java/com/youtube/vitess/client/Proto.java#L228)):*
Done.

---

*[java/client/src/main/java/com/youtube/vitess/client/Proto.java, line 231 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1495#-KAhvODK6P7tr89-EAL1-r1-231:-KCdlm88aTRQF-9Joo9T:-499146009) ([raw file](https://github.com/youtube/vitess/blob/a8136b8bfe6ac7914062989903ea9e3667b65e19/java/client/src/main/java/com/youtube/vitess/client/Proto.java#L231)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1495)*
<!-- Sent from Reviewable.io -->

enisoc commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[java/client/src/main/java/com/youtube/vitess/client/Proto.java, line 217 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1495#-KCqjY4nerBCqSkiosAv:-KCqjY4nerBCqSkiosAw:163450076) ([raw file](https://github.com/youtube/vitess/blob/0db7ee2b197a19c44740db4a34247e78225706af/java/client/src/main/java/com/youtube/vitess/client/Proto.java#L217)):*
Is there a benefit to using `String.valueOf(1)` instead of just `"1"`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1495)*
<!-- Sent from Reviewable.io -->

harshit-gangal commented:
*[java/client/src/main/java/com/youtube/vitess/client/Proto.java, line 217 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1495#-KCqjY4nerBCqSkiosAv:-KCr7LOgmpHRaK41OKSU:1466600515) ([raw file](https://github.com/youtube/vitess/blob/0db7ee2b197a19c44740db4a34247e78225706af/java/client/src/main/java/com/youtube/vitess/client/Proto.java#L217)):*
No Particular reason. Change Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1495)*
<!-- Sent from Reviewable.io -->

enisoc commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Looks ready to merge, once CLA is approved.

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1495#-:-KCrGSzstLfKVzzq7YNT:2002884425)*
<!-- Sent from Reviewable.io -->

harshit-gangal commented:
Needs to approve pull request here
https://pullapprove.com/youtube/vitess/pull-request/1495/
